### DISCUSSION TITLE
What is the recommended architecture for utilizing fleet/osquery functionality?

### DISCUSSION THREAD

**Discussion by sam-acker** [Security-Onion-Solutions/securityonion#5512] [Unsupported Versions] [answered]
Hello,

Have run into some trouble getting osquery/fleet set up for a particular use case that I have. First, let me describe what I'd like to achieve:

Trying to build out a model of a managed security service provider where we have a security onion manager node and search node on our network and a hypervisor with a security onion forward node (and some other services) deployed to multiple clients. 

Using the standard distributed architecture with fleet running on the manager I was easily able to get "on network" (in the same subnet as the manager) osquery endpoints enrolled and functional. I am now however having some trouble getting "off network" (on the client's network) endpoints enrolled.

Here are the steps I've taken so far:
- added NAT rule on our firewall so that machines on the client network can reach the manager node on port 8090 at the edge router's ip
-  have confirmed that this rule works by listening on the manager with tcpdump and sending packets from the client network with netcat
- run `so-allow` to allow osquery endpoints on the client range
- regenerated osquery install packages with `sudo salt-call state.apply fleet.event_gen-packages`
- enrollment with these packages failed as they don't know to point to the NAT'ed ip of the manager (the WAN address of the router on port 8090)
- tried to enroll with the package on mac (changed the address and port in the `launcher.flags` file along with the other steps recommended in the documentation)
- tried to enroll the manual way on Ubuntu following the instructions in the "add new host" button on fleet (changed server ip and port in the flagfile, used the server certificate and secret they provide, run with `osqueryd --flagfile=flagfile.txt --verbose`) throws a __"certificate verify failed"__ error (see below)
- When using `curl -v -X POST` to troubleshoot it throws an __"unable to get local issuer certificate"__ error (see below) (should I maybe use the cert at `/etc/ssl/certs/intca.crt` instead of the one fleet provides as `fleet.pem`? Or `/etc/so-launcher/roots.pem`?)

Not sure what to try next or if I'm even on the right track/using the right architecture for what I'm trying to achieve. Would really appreciate some guidance here. Thanks!

### curl error
<img width="533" alt="curl_error" src="https://user-images.githubusercontent.com/37353790/133340817-da33121c-1a46-4a6e-8c44-007eb437f3f7.png">

### osqueryd error
<img width="496" alt="osqueryd_error" src="https://user-images.githubusercontent.com/37353790/133340820-63488cfb-353b-4599-b9f8-588462da20dc.png">

**Comment by sam-acker** [comment]
Any updates on this? Read this [excellent article](https://defensivedepth.com/2020/04/02/kolide-fleet-breaking-out-the-osquery-api-web-ui/) by @defensivedepth and I think I'm understanding how this works a little better in that the agents don't directly enroll to the fleet server on the manager but rather they hit 8090 on the nginx container first and then the traffic gets forwarded. My ssl knowledge is severely lacking but I think this probably has something to do with the cert errors I'm getting. Tried using the cert at `/etc/pki/nginx/server.crt` found on the `so-nginx` container in the agent enrollment process but still no luck.

**Comment by defensivedepth** [comment, marked as answer]
How is the "off-network" SO node connected to your Manager? Are you seeing any connectivity issues with it in general?

**Reply by sam-acker** [reply]
Yep, the off-network SO forward node is connected to the manager over a site to site vpn. There doesn't seem to be any connection problems there. The problem is that we will not be able to send any client endpoint data directly over the tunnel to the Manager node. We can send data from endpoints to the Forward node and then have the forward node send the data over the tunnel to the Manager (the way Wazuh does it), but it doesn't seem like Fleet is designed this way. What we're trying to do to work around this is to send the osquery traffic from the endpoint to the edge router of the network the manager is on and use a NAT rule to pass that traffic back to the Manager. Please feel free to suggest a different setup than this or tell me if I've misunderstood how osquery/fleet traffic is suppossed to flow! Can also draw a net diagram for our current setup if that would help.

**Reply by defensivedepth** [reply]
I will need to think on this more, I'm thinking the issue is the cert. A net diagram would definitely help lol!

**Reply by sam-acker** [reply]
![Model Client Net Diagram](https://user-images.githubusercontent.com/37353790/134751193-73b914d5-6e5a-4323-bc1a-23bdb879b17f.jpeg)

Hope this helps and is clear, please let me know if you have any questions! Just a note, the purple site to site VPN is going over the orange VPN between the Lab router and the Model Client Router, just couldn't draw it that way very well lol. Also to clarify, that orange VPN is the "fake internet" meant to simulate a connection over the internet (testing this all on a LAN so the VPN between the Lab and the Model client is simply meant to simulate a remote connection between them as it would be in a real situation).

**Reply by defensivedepth** [reply]
So the osquery endpoints are trying to connect to the Manager which has a NAT'ted IP? Does the Forward Node connect to the Manager this way as well?

**Reply by sam-acker** [reply]
Yes, exactly. The forward node talks directly to the management interface of the manager (the one set up during the initial install) over that purple site to site vpn that goes between the two virtual routers. Have tested that client endpoint machines can send traffic to the WAN ip of the security service provider vRouter on port 8090 and the manager hears that traffic due to the NAT rule.

**Reply by sam-acker** [reply]
Any updates on this?

**Reply by defensivedepth** [reply]
I believe the issue is that the cert check is failing because it doesn't have the NAT IP as part of the SAN, only the internal IP. I would suggest opening up an issue in this repo and referencing this discussion --> Feature Request: Allow osquery clients to connect via NAT.

**Comment by EricHStevens** [comment]
Hi @defensivedepth,
We built a new configuration in which we have Fleet on the Model Client. We then added a host on the same subnet and got the same certificate error:
![image_fleet_error_2](https://user-images.githubusercontent.com/15239203/138112401-79d60c7b-c8a1-4c65-a7df-4a97928e0900.png)

Thus, I think we have proven that the problem is not with the NAT IP.

Is it possible Fleet is trying to verify the certificate against the wrong certificate?
![image](https://user-images.githubusercontent.com/15239203/138113110-aadf441b-4c30-4b6c-b354-3f072a4fcc0e.png)
Any idea on how to verify this?

Is it possible https://192.168.1.120:8080/api/v1/osquery/enroll is not the correct path?

I am really just looking for anything to debug.

**Reply by defensivedepth** [reply]
Can you please send me the following output (sanitized as needed):

From the Manager:

`sudo salt-call pillar.get global`

`sudo cat /opt/so/saltstack/local/salt/fleet/packages/launcher.flags`

**Reply by EricHStevens** [reply]
$ sudo salt-call pillar.get global
local:
    ----------
    airgap:
        False
    cortexkey:
        FD7t9tqdIx1tEAOhsq8t
    cortexorgname:
        SecurityOnion
    cortexorguserkey:
        fWIcf2GBvgE9GcHakLdN
    cortexplaysecret:
        9rwXWt06yAD3absWgCKh
    dockernet:
        172.17.0.0
    fleet_custom_hostname:
        None
    fleet_hostname:
        so-fleet
    fleet_ip:
        10.1.0.25
    fleet_manager:
        True
    fleet_node:
        True
    fleet_packages-timestamp:
        2021-11-09-21:21
    fleet_packages-version:
        3
    hivekey:
        rHLuZMnbXmOpf28z4C6W
    hiveplaysecret:
        peQMK0hEzBPgJD7vAFak
    hnmanager:
        10.0.2.0/24,10.1.0.0/24
    ids:
        Suricata
    imagerepo:
        security-onion-solutions
    managerip:
        10.0.2.134
    mdengine:
        ZEEK
    pipeline:
        redis
    sensoronikey:
        E4UNCtONohBT9tqVbneK
    soversion:
        2.3.80
    url_base:
        10.0.2.134
    wazuh:
        1

**Reply by EricHStevens** [reply]
$ sudo cat /opt/so/saltstack/local/salt/fleet/packages/launcher.flags
enroll_secret_path /etc/so-launcher/secret
hostname so-fleet:8090
root_directory /var/so-launcher/securityonion
osqueryd_path /usr/local/so-launcher/bin/osqueryd
root_pem /etc/so-launcher/roots.pem
autoupdate
update_channel stable

**Reply by EricHStevens** [reply]
In this configuration we put the so-manager in AWS.
There is a tunnel connecting the AWS VPC environment with the Forward Sensor and Fleet.
I can confirm that fleet is on IP 10.1.0.25 and the host name is so-fleet. However, I should note that the DNS does not resolve so-fleet to 10.1.0.25 on so-manager (ip-10-0-2-134).

Do you think this will be a problem? That one cannot use the hostname so-fleet, but need to use the ip 10.1.0.25?
From the manager node:
```
$ sudo ping so-fleet
ping: so-fleet: Temporary failure in name resolution
```
Another item that might be helpful. During installation of the fleet node I had to do it twice.
The first time I named the fleet node 'so-fleet-std_fleet' and it failed. I have attached the log file:
[sosetup_fleet_1st.log](https://github.com/Security-Onion-Solutions/securityonion/files/7508620/sosetup_fleet_1st.log)
I then tried to install a 2nd time, where I named the node so-fleet. This seemed to work and everything installed fine:
[sosetup_fleet_2nd.log](https://github.com/Security-Onion-Solutions/securityonion/files/7508634/sosetup_fleet_2nd.log)

I can see the fleet node in the grid (third item):
![fleet_in_grid](https://user-images.githubusercontent.com/15239203/141022821-4fe977fc-dfe5-4516-9992-21b6c13dac2b.png)
:
However, the FleetDM instance looks like it is the old one on the so-manager:
![fleet_not_sure_the_right_one](https://user-images.githubusercontent.com/15239203/141022891-a8102a89-201c-46ed-9fed-31cbc15dd46a.png)

**Reply by EricHStevens** [reply]
I know this took a long time, I was trying to get so-manager into AWS, before I could get you the salt and cat commands.

**Reply by EricHStevens** [reply]
I copied the launcher.deb from the remote fleet instance, located at /opt/so/conf/fleet/packages/launcher.deb, to a VM attached to the networked and tried to install it, and it didn't appear to work:
```
$ sudo dpkg -i /home/ubuntu/Downloads/launcher.deb
[sudo] password for ubuntu: 
(Reading database ... 127270 files and directories currently installed.)
Preparing to unpack .../ubuntu/Downloads/launcher.deb ...
Unpacking launcher-final (3.1.1) over (3.1.1) ...
Setting up launcher-final (3.1.1) ...
```

Doesn't look like it installed correctly. I did add the IP of this VM to the so-manager using so-allow.

**Reply by defensivedepth** [reply]
A couple things:

- When you attempt to login manually to the Standalone Fleet web interface, what happens? `https://10.1.0.25/fleet`

- Make sure that the Debian VM you attached to the network can resolve `so-fleet`. Then run the following and post any error logs that you see. Focus on connection errors.  
 
`sudo /usr/local/so-launcher/bin/launcher -config /etc/so-launcher/launcher.flags --debug`

**Reply by EricHStevens** [reply]
When trying to connect to https://10.1.0.25/fleet I get the following security error:
![image](https://user-images.githubusercontent.com/15239203/141197023-3e3bcfd4-0583-4d40-a660-4759272d10c0.png)
However, I can still login into the security onion analyst page https://10.0.2.134.

The VM I am trying to install the "launcher.deb" file on cannot see so-fleet (I do not have a DNS server running in the environment). So, I just added so-fleet 10.1.0.25 to /etc/hosts. I then ran the command:
sudo /usr/local/so-launcher/bin/launcher -config /etc/so-launcher/launcher.flags --debug
![image](https://user-images.githubusercontent.com/15239203/141198133-9486ed21-1801-4776-8b15-f43552502179.png)

**Reply by EricHStevens** [reply]
Tried running this command:
$ sudo salt-call state.highstate
local:
    Data failed to compile:
----------
    Detected conflicting IDs, SLS IDs need to be globally unique.
    The conflicting ID is 'so-fleet' and is found in SLS 'base:fleet' and SLS 'base:fleet.install_package'

Which seems to imply the manager knows it is having a problem with fleet.

**Comment by EricHStevens** [comment]
Effectively, we did a fresh install of Security Onion 2.3.80 for both the manager, fleet, made sure to do an so-allow, and setup a firewall where the original IP was visible on both sides of the tunnel and things worked.

I believe we can consider this question answered and closed.
